### PR TITLE
Fix KVO crashes

### PR DESCRIPTION
**Pull request by tbodt**

### CODE REVIEW

File: app/NSObject+SaneKVO.m [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    KVOBlock newBlock = ^{
+        id owner = weakOwner;
+        if (owner) {
+            block(owner);
+        }
+    };
+    @synchronized (owner) {
+        for (NSString *keyPath in keyPaths) {
+            NSMutableSet *observations = objc_getAssociatedObject(owner, kKVOObservations);
+            if (observations == nil) {
```
Comment (saagarjha) [inline app/NSObject+SaneKVO.m:36]: `!observations`

File: app/NSObject+SaneKVO.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef void (^KVOBlock)(void);
+
+@interface KVOObservation : NSObject {
+    BOOL _enabled;
+    __weak id _object;
```
Comment (saagarjha) [inline app/NSObject+SaneKVO.h:16]: Possibly `id<NSKeyValueObserving>`?

File: app/NSObject+SaneKVO.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+//
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef void (^KVOBlock)(void);
+
+@interface KVOObservation : NSObject {
+    BOOL _enabled;
```
Comment (saagarjha) [inline app/NSObject+SaneKVO.h:15]: Is this used?

File: app/NSObject+SaneKVO.m [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    return observation;
+}
+
+- (void)observe:(NSArray<NSString *> *)keyPaths options:(NSKeyValueObservingOptions)options owner:(id)owner usingBlock:(void (^)(id self))block {
+    __weak id weakOwner = owner;
+    KVOBlock newBlock = ^{
+        id owner = weakOwner;
+        if (owner) {
+            block(owner);
+        }
```
Comment (saagarjha) [inline app/NSObject+SaneKVO.m:31]: This sounds like a good time to disable early

File: app/Roots.m [hunk 5]
```diff
@@ -53,18 +53,21 @@ - (instancetype)init {
         [self addObserver:self forKeyPath:@"roots" options:0 context:nil];
```
Comment (saagarjha) [inline app/Roots.m:53]: Can we fix this?

**Review by saagarjha — COMMENTED** [review]
Initial comments

File: app/NSObject+SaneKVO.h (hunk 2 above)
Comment (saagarjha) [reply app/NSObject+SaneKVO.h:16]: Nevermind

File: app/NSObject+SaneKVO.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+@interface NSObject (SaneKVO)
+
+- (KVOObservation *)observe:(NSString *)keyPath
+                    options:(NSKeyValueObservingOptions)options
+                 usingBlock:(KVOBlock)block;
+- (void)observe:(NSArray<NSString *> *)keyPaths
+        options:(NSKeyValueObservingOptions)options
+          owner:(id)owner
+     usingBlock:(void (^)(id self))block;
```
Comment (saagarjha) [inline app/NSObject+SaneKVO.h:31]: These should be prefixed

File: app/NSObject+SaneKVO.m (hunk 4 above)
Comment (saagarjha) [reply app/NSObject+SaneKVO.m:31]: Actually, I would assert here, this should never happen…

File: app/NSObject+SaneKVO.h (hunk 3 above)
Comment (tbodt) [reply app/NSObject+SaneKVO.h:15]: Yes, in -disable

File: app/NSObject+SaneKVO.h (hunk 6 above)
Comment (tbodt) [reply app/NSObject+SaneKVO.h:31]: but I don't wanna

File: app/NSObject+SaneKVO.m (hunk 1 above)
Comment (tbodt) [reply app/NSObject+SaneKVO.m:36]: I like this better, also it doesn't really matter

File: app/Roots.m (hunk 5 above)
Comment (tbodt) [reply app/Roots.m:53]: Sure

File: app/NSObject+SaneKVO.m (hunk 4 above)
Comment (tbodt) [reply app/NSObject+SaneKVO.m:31]: Done

File: app/NSObject+SaneKVO.h (hunk 6 above)
Comment (saagarjha) [reply app/NSObject+SaneKVO.h:31]: But you should.

File: app/NSObject+SaneKVO.h (hunk 3 above)
Comment (saagarjha) [reply app/NSObject+SaneKVO.h:15]: My actual concern is that this never seems to be set to `YES`

Comment (tbodt) [reply app/NSObject+SaneKVO.h:15]: There needs to be some way to mark this as disabled so it won't be disabled a second time. Also maybe someday I'll add -enable.

File: app/NSObject+SaneKVO.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+//  iSH
+//
+//  Created by Theodore Dubois on 11/10/20.
+//
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef void (^KVOBlock)(void);
```
Comment (saagarjha) [inline app/NSObject+SaneKVO.h:12]: Does this really need a `typedef`

File: app/NSObject+SaneKVO.h (hunk 3 above)
Comment (saagarjha) [reply app/NSObject+SaneKVO.h:15]: My concern is that `-disable` seems to be a no-op at the moment

Comment (tbodt) [reply app/NSObject+SaneKVO.h:15]: Oops

Comment (tbodt) [reply app/NSObject+SaneKVO.h:15]: Oops

File: app/NSObject+SaneKVO.h (hunk 7 above)
Comment (tbodt) [reply app/NSObject+SaneKVO.h:12]: Not really

File: app/NSObject+SaneKVO.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+@interface NSObject (SaneKVO)
+
+- (KVOObservation *)observe:(NSString *)keyPath
+                    options:(NSKeyValueObservingOptions)options
+                 usingBlock:(void (^)(void))block;
+- (void)observe:(NSArray<NSString *> *)keyPaths
+        options:(NSKeyValueObservingOptions)options
+          owner:(id)owner
+     usingBlock:(void (^)(id self))block;
```
Comment (saagarjha) [inline app/NSObject+SaneKVO.h:29]: `owner`

File: app/Roots.m [hunk 9]
```diff
@@ -50,21 +51,28 @@ - (instancetype)init {
...
+}
+- (void)setDefaultRoot:(NSString *)defaultRoot {
+    [NSUserDefaults.standardUserDefaults setObject:defaultRoot forKey:kDefaultRoot];
+}
+
 - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
-    if ([keyPath isEqualToString:@"defaultRoot"]) {
-        [NSUserDefaults.standardUserDefaults setObject:self.defaultRoot forKey:kDefaultRoot];
-    } else if ([keyPath isEqualToString:@"roots"]) {
+    if ([keyPath isEqualToString:@"roots"]) {
```
Comment (saagarjha) [inline app/Roots.m:75]: Can we delete this code?

**Review by saagarjha — COMMENTED** [review]
Last couple things I noticed.

File: app/Roots.m (hunk 9 above)
Comment (tbodt) [reply app/Roots.m:75]: Yes

File: app/NSObject+SaneKVO.h (hunk 8 above)
Comment (tbodt) [reply app/NSObject+SaneKVO.h:29]: The argument is self because it is typical for the caller to want to name it self

Comment (saagarjha) [reply app/NSObject+SaneKVO.h:29]: You're welcome to call it `self` if that is the thing you are setting as the owner, but the header should at least match the semantics of the code (which is generic).

Comment (tbodt) [reply app/NSObject+SaneKVO.h]: Changed to leave out the argument name

**Pull request merged**